pastas_plugins/pest/solver.py

```diff
@@ -0,0 +1,1159 @@
+            se = (
+                pyemu.ObservationEnsemble.from_csv(
+                    pst=pst, filename=self.master_ws / f"pest.{iteration}.obs.csv"
+                )
+                .transpose()
+                .set_index(self.ml.observations().index)
+            )
+        else:
+            ipar = self.parameter_ensemble(iteration=iteration).transpose()
+            ipar.index = self.ml.parameters.index[self.vary]
+
+            tmin = self.ml.settings["tmin"] if tmin is None else pd.Timestamp(tmin)
+            tmax = self.ml.settings["tmax"] if tmax is None else pd.Timestamp(tmax)
+            freq = (
+                "D"
+                if self.ml.settings["freq"] is not None
+                else self.ml.settings["freq"]
+            )
+            se = pd.DataFrame(
+                np.nan,
+                columns=ipar.columns,
+                index=pd.date_range(start=tmin, end=tmax, freq=freq),
+            )
+
+            for idx in ipar.columns:
+                self.ml.parameters.loc[ipar.index, "optimal"] = ipar.loc[:, idx].values
+                se.loc[:, idx] = (
+                    self.ml.simulate(tmin=tmin, tmax=tmax).loc[se.index].values
+                )
+
+        return se
+
+    def observation_ensemble(self) -> pyemu.ObservationEnsemble:
+        """
+        Generate an observation ensemble from a CSV file. This method reads a
+        PEST control file and a corresponding observation noise CSV file to
+        create an observation ensemble.
+
+        Returns
+        -------
+        pyemu.ObservationEnsemble
+            The generated observation ensemble.
+        """
+
+        pst = pyemu.Pst(str(self.master_ws / "pest.pst"))
+        oe = pyemu.ObservationEnsemble.from_csv(
+            pst=pst, filename=self.master_ws / "pest.obs+noise.csv"
+        )
+        return oe
+
+    def jacobian(self, iteration: int = 0) -> pd.DataFrame:
+        """
+        Calculate the Jacobian matrix for the given iteration.
+        The Jacobian matrix is computed using the simulation ensemble and
+        parameter ensemble for the specified iteration. The calculation
+        involves normalizing the ensembles and using the pseudo-inverse
+        of the parameter ensemble.
+
+        Parameters:
+        -----------
+        iteration : int, optional
+            The iteration number for which the Jacobian is calculated.
+            Default is 0.
+
+        Returns:
+        --------
+        pd.DataFrame
+            A DataFrame representing the Jacobian matrix, with the same
+            indices as the observation ensemble and the same columns as
+            the parameter ensemble.
+        """
+        # jac_ies needs to be calculated manually
+        obs_ies = self.simulation_ensemble(
+            iteration=iteration, from_file=True
+        ).transpose()
+        par_ies = self.parameter_ensemble(iteration=iteration)
+        jac = PestIesSolver.jacobian_emperical(obs_ies.values, par_ies.values)
```

Comment:
Using the Pest implementation and not the local one?

Reply:
Not really sure what you mean? But pestpp-ies does not store the jacobian for each time step. So it has to be calculated from the simulated value and parameter ensembles.

Reply:
My bad, I didn't realize it was referring the the method below, I thought you were using some PyEMU class method. Why not use self?

Reply:
Because the `jacobian_emperical` is a `@staticmethod`. I use this sometimes standalone. same as `parameter_distribution` and `generate_observation_noise`

Reply:
You can still use `self` to access the staticmethod, but it doesn't really matter. I should have read better :see_no_evil: .